### PR TITLE
thread local storage checking

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Implement our own simple macros for checking for thread local storage
support from the compiler.

## Expected Performance Changes

none

## Known Issues

none

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
There is an issue with Intel compilers on OS/X using thread local storage, but it exists prior to this change, as well. I can confirm this change detects and uses TLS correctly on Linux and OS/X with gcc and clang compilers.

**Comment by raffenet** [comment]
Actually hold off on review. I have another idea. Patch incoming...

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

File: src/mpl/configure.ac [hunk 1]
```diff
@@ -684,7 +684,31 @@ AC_DEFINE_UNQUOTED([THREAD_PACKAGE_NAME],[$THREAD_PACKAGE_NAME],[set to the name
...
+    AC_MSG_RESULT([$pac_cv_have__thread_local])
+    AC_MSG_CHECKING([for __thread])
+    AC_LINK_IFELSE(
+        [AC_LANG_PROGRAM([__thread int foo = 0;],[foo = 1])],
+	[pac_cv_have___thread=yes],[pac_cv_have___thread=no])
+    AC_MSG_RESULT([$pac_cv_have___thread])
+    AC_MSG_CHECKING([for __declspec(thread)])
+    AC_LINK_IFELSE(
+        [AC_LANG_PROGRAM([__declspec(thread) int foo = 0;],[foo = 1])],
+	[pac_cv_have___declspec_thread]=yes,[pac_cv_have___declspec_thread=no])
```
Comment (hzhou) [inline src/mpl/configure.ac]: Misplaced closing `]`?

**Comment by hzhou** [comment]
test:mpich/warnings/more

**Comment by hzhou** [comment]
test:mpich/warnings/more

**Comment by hzhou** [comment]
test:mpich/warnings/more

File: src/mpl/configure.ac (hunk 1 above)
Comment (hzhou) [reply src/mpl/configure.ac]: Since we are only going to use one TLS definition, can we do `elif` or `if -z $pac_cv_tls ...` (assuming we'll set pac_cv_tls to the first keyword that's available)?

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:icc/warnings

**Comment by hzhou** [comment]
test:more/warnings

File: src/mpl/configure.ac (hunk 1 above)
Comment (raffenet) [reply src/mpl/configure.ac]: I initially tried nesting the `AC_LINK_IFELSE` macros so they would only run if the previous one failed, but it was kind of hard to follow. I'm open to changing the logic here, but I'm not sure I'm getting the suggestion. Can you put it in pseudo-code? Or just add a commit to this branch?

Comment (raffenet) [reply src/mpl/configure.ac]: Hmm, maybe the nesting isn't too bad. New patch coming.

File: src/mpl/configure.ac [hunk 2]
```diff
@@ -684,7 +684,22 @@ AC_DEFINE_UNQUOTED([THREAD_PACKAGE_NAME],[$THREAD_PACKAGE_NAME],[set to the name
...
+    [AC_MSG_RESULT([yes])
+    AC_DEFINE([TLS],[__thread],[Using __thread for TLS])],
+    [AC_MSG_RESULT([no])
+    AC_MSG_CHECKING([for __declspec(thread)])
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([__declspec(thread) int foo = 0;],[foo = 1])],
+    [AC_MSG_RESULT([yes])
+    AC_DEFINE([TLS],[__declspec(thread)],[Using __declspec(thread) for TLS])],
+    [AC_MSG_RESULT([no])
+    AC_MSG_WARN([Compiler does not support thread local storage])])])])
 fi
```
Comment (hzhou) [inline src/mpl/configure.ac]: Reads just like LISP :)

What I was having in mind is:
```
AC_MSG_CHECKING([for thread-local storage])
if test -z $pac_cv_tls ; then
    AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([_Thread_local int foo=0;],[foo=1])],
            [pac_cv_tls=_Thread_local], []
     )
fi
if test -z $pac_cv_tls ; then 
    AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([__thread int foo=0;],[foo=1])],
            [pac_cv_tls=__thread], []
     )
fi
if test -z $pac_cv_tls ; then 
    AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([__declspec(thread) int foo=0;],[foo=1])],
            [pac_cv_tls="__declspec(thread)"], []
     )
fi
if -n $pac_cv_tls ; then
    AC_MSG_RESULT([no])
    # AC_MSG_WARN  or ERROR
else
    AC_MSG_RESULT([$pac_cv_tls])
    AC_DEFINE([TLS], [$pac_cv_tls], [Defined as the attribute keyword for thread-local storage.])
fi
```

Either way, this is a handful to read for a single test. Can we wrap it in `AC_DEFUN([PAC_CC_CHECK_TLS],...` in `confdb/aclocal_cc.m4`? 

PS: why do we need the `extern` for `_Thread_local` test?

Comment (raffenet) [reply src/mpl/configure.ac]: The usage in MPICH is with `extern`, so the test closely matches. I also used `AC_LINK_IFELSE` because at least one toolchain had compiler support, but broke down at the link step.

Comment (raffenet) [reply src/mpl/configure.ac]: The other suggestions here look good to me. I'll push a new version.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

File: src/mpl/configure.ac (hunk 2 above)
Comment (raffenet) [reply src/mpl/configure.ac]: Actually it looks like the link test is sufficient for what we need. `extern` doesn't affect the results, so I'm OK to remove it.

File: confdb/aclocal_cc.m4 [hunk 3]
```diff
@@ -1651,3 +1651,31 @@ AC_DEFUN([PAC_C_STATIC_ASSERT], [
...
+        AC_LINK_IFELSE([AC_LANG_PROGRAM([_Thread_local int foo=0;],[foo=1])],
+            [pac_cv_tls=_Thread_local])
+    fi
+    if test -z $pac_cv_tls ; then
+        AC_LINK_IFELSE( [AC_LANG_PROGRAM([__thread int foo=0;],[foo=1])],
+            [pac_cv_tls=__thread])
+    fi
+    if test -z $pac_cv_tls ; then
+        AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([__declspec(thread) int foo=0;],[foo=1])],
+            [pac_cv_tls="__declspec(thread)"])
```
Comment (hzhou) [inline confdb/aclocal_cc.m4]: Is there a reason to use `AC_COMPILE_IFELSE` here? Actually, I am curious about which toolchain had problems with linking and wonder what causes it.

File: confdb/aclocal_cc.m4 [hunk 4]
```diff
@@ -1651,3 +1651,31 @@ AC_DEFUN([PAC_C_STATIC_ASSERT], [
...
 ])
+
+dnl
+dnl PAC_CC_CHECK_TLS - Test for thread local storage support
+dnl
+dnl will AC_DEFINE([TLS]) to a compiler supported TLS keyword
+dnl
+AC_DEFUN([PAC_CC_CHECK_TLS], [
+    AC_MSG_CHECKING([for thread-local storage])
+    if test -z $pac_cv_tls ; then
```
Comment (hzhou) [inline confdb/aclocal_cc.m4]: Maybe we should use `AC_CACHE_CHECK` here?

File: confdb/aclocal_cc.m4 (hunk 3 above)
Comment (raffenet) [reply confdb/aclocal_cc.m4]: This was a mistake. I am fixing it. The problematic toolchain is `icc` on macOS.

File: confdb/aclocal_cc.m4 (hunk 4 above)
Comment (raffenet) [reply confdb/aclocal_cc.m4]: Done.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
@hzhou I added fixes for your last round of comments. Tests look OK.
